src/geoFire/geoQuery.ts

```diff
@@ -0,0 +1,526 @@
+import * as firebase from 'firebase';
+
+import { GeoCallbackRegistration } from './geoCallbackRegistration';
+import { decodeGeoFireObject, distance, encodeGeohash, geoFireGetKey, geohashQueries, validateCriteria, validateLocation } from './geoFireUtils';
+
+import { QueryCriteria } from './interfaces';
+
+/**
+ * Creates a GeoQuery instance.
+ */
+export class GeoQuery {
+  // Event callbacks
+  private _callbacks: any = { ready: [], key_entered: [], key_exited: [], key_moved: [] };
+  // Variable to track when the query is cancelled
+  private _cancelled: boolean = false;
+  private _center: number[];
+  // A dictionary of geohash queries which currently have an active callbacks
+  private _currentGeohashesQueried: any = {};
```

Comment:
We should probably provide an interface here as well. This should really be a map of string to the object [here](https://github.com/firebase/geofire-js/pull/169/files#diff-63a1f564d018e28953019ef5f97418f8R429) no?